### DISCUSSION TITLE
How to (un)lock the Database connection?

### DISCUSSION THREAD

**Discussion by dusadrian** [duckdb/duckdb#10397] [General]
Hi all,

I am building an Electron app (under Node.js) that will lock the connection to a DuckDB file for reading and writing.
At the same time, I would like to have another program (say R) that would use the data in read_only mode to create reports. However, R cannot connect to the database file because it is being locked by the Electron app.

Is there a way to allow another software to (only) read a locked DuckDB file?
Alternatively, is there a way to close() the connection from the Node.js app, to allow R to read the data?

I already spent an entire day reading the documentation and trying different setups, with no avail.
I use something like:

```
const db = new DuckDB.Database(dbFile);
```

and after each async SQL query I try to close the connection using:
```
db.close()
```

but the lock refuses to be released.
Something similar happens using the SQLTools extension for VS Code, I can create a connection to that file but, despite disconnecting it, the lock does not disappear unless quitting VS Code.

Any hint would be greatly appreciated, thanks in advance.

**Comment by Tishj** [comment]
New release will contain this, which should help debugging situations like this

https://github.com/duckdb/duckdb/pull/9988

It also sounds like the node client might have a bug if the lock isnt released with close()

**Reply by dusadrian** [reply]
Excellent, any idea when will the new version be released?
About the node client bug, where should I file a report?

**Reply by Tishj** [reply]
I believe the next release will be in 3 weeks, but that's still a little up in the air.
I don't know for sure if that's a bug, it could just be that the OS is too slow to release the filesystem lock, but if it's never released then maybe. Probably here <https://github.com/duckdb/duckdb-node>

**Reply by dusadrian** [reply]
Thanks for the heads up. That might be, I am temporarily working with an old MacBook Air with Catalina (still), so that might be an issue. As soon as my main computer will arrive, I will test on Windows and a newer MacOS to see if this is still the case.
For the moment, at the first run the lock is released in about 30 seconds (which I read is the default for http_timeout but that might be a coincidence). At the next run of an SQL query, the lock is sometimes released immediately but most often never, so it might very well be an OS issue.

**Reply by Mause** [reply]
Do make sure you're actually waiting for the `close()` callback to complete as well

**Reply by dusadrian** [reply]
I've timed it, sometimes waiting for ~20 minutes just for the sake of it, and still no luck.
If there is there any other way to make sure the close() callback is completed, that would be nice.

**Reply by Mause** [reply]
You can try `FORCE CHECKPOINT` first, to see if that's the issue, before closing

**Reply by dusadrian** [reply]
I ran `db.run("FORCE CHECKPOINT")` before `db.close()`
(in hoping this is how to do it), but still the same behaviour. It may be an OS issue, will test on other machines and return with a report.

**Reply by Mause** [reply]
Do remember to actually wait for the checkpoint to complete as well. If you are able to provide a reproduction, that would be of great help

**Reply by dusadrian** [reply]
Thanks @Mause, I now have tried on three different computers. Must be doing something wrong, because I don't seem to get the expected results anywhere.
I've tried on a macOS Catalina with an Intel chipset, as well as on macOS Sonoma with an M2 Silicon processor. I also tried on a virtual machine on Windows 11.

This is a link to a GitHub repository called [ducktest](https://github.com/dusadrian/ducktest/), for reproduction. It needs Node.js (I installed the latest version), and then `npm install` should create the required node_modules, after which `nom start` should start the app.

It consists of a single button "Push me", which at click it fires an event that queries the duckdb database for the first 10 lines from the table "df".

The database connection is made at the file [database.ts](https://github.com/dusadrian/ducktest/blob/main/src/database/database.ts) in lines 35-37. My intention is to close the connection after the query, to be able to read the test.duckdb with R. I probably don't do it right, but this is attempted in lines 53-54, where I also FORCE CHECKPOINT.

The .wal file is created after the first button push, and it disappears after ~30 seconds. It I push it multiple times, the .wal file won't disappear (not even after quitting the application). If quitting, the database is freed from the lock (despite the .wal file not being erased).

Under Windows 11, the second button push runs into: `Connection Error: Connection was never established or has been closed already`.

I would be grateful for any further hint,
Adrian

**Reply by carlopi** [reply]
Can you provide a callback to close, and await on the result of that callback? [ Using a similar pattern as the callback provided to db.all ]

**Reply by dusadrian** [reply]
Just pushed to the repo, same result unfortunately...

**Reply by carlopi** [reply]
I meant something like:
```js
        const closeDB = new Promise<Boolean>((resolve) => {
            db.close((err,res) => { if (err) {console.log(err);} console.log("closed?"); resolve(true);} );
        })
        await closeDB;
```

**Reply by carlopi** [reply]
It might not be super intuitive, but any duckdb function basically takes a callback, and so needs to be awaited on the callback result.

Project async-duckdb is an alternative API that might be worth taking a look at, I believe it provides Promise-wrappings out of the box.

**Reply by dusadrian** [reply]
Just did (and pushed), but the end result doesn't change. The first click closes the connection after 30 seconds. Beyond that, a second click does not close the connection (and the .wal file does not get deleted).

**Reply by dusadrian** [reply]
As far as I can see, the connection does get closed:
`Database {default_connection: null}`
but the lock is not removed. This probably happens because the object `db` is not garbage collected.

In R, I do `dbDisconnect(con, shutdown = TRUE)`, which I am not sure (and I did not find in the documentation) how to achieve in Node.js

The `shutdown = TRUE` is meant precisely for garbage collect.

**Reply by Mause** [reply]
Can you commit your package lock file? Perhaps I'm not using the same versions as you:

![image](https://github.com/duckdb/duckdb/assets/1405026/f8e2349c-f2b6-485f-8672-d8f7abd1a6f1)

**Reply by Mause** [reply]
And I'm able to open the database file using the cli without issue

**Reply by dusadrian** [reply]
Just committed the package-lock.
After 30 seconds from the first button click, I am also able to open the database using the cli. The problem is pressing it a second (third etc) time, but after the first lock being released.
Procedure:
- start the application
- press the button
- wait 30 seconds (the lock will be removed)
- press the button a second time

Now the lock will not be removed anymore, and the database will not be opened from the cli.

**Reply by dusadrian** [reply]
Hello @Mause, @Tishj,
I see the new version 0.10 was released, and wanted to ask if this Node.js behavior was perhaps handled.
The NPM version is still 0.9.2, is there a roadmap to update that client API?

**Reply by szarnyasg** [reply]
Yes, we'll update the Node client in the coming days. For the upcoming patch releases, see https://duckdb.org/dev/release-calendar.html

**Comment by dusadrian** [comment]
> New release will contain this, which should help debugging situations like this
>
> It also sounds like the node client might have a bug if the lock isnt released with close()

Hi @Tishj, @szarnyasg,

Thank you very much for the new version 0.10 as well as the Node client. I just tried to close the database connection after one simple query, and the behavior is still the same (the connection remains locked) even when using `db.close()`
(please see the MRE from [https://github.com/dusadrian/ducktest/](https://github.com/dusadrian/ducktest/)

Am I perhaps using `db.close()` incorrectly?

From the concurrency page, I read that DuckDB supports either a single connection with read and write, or multiple connections in read only more. What I need is a connection to read and write, plus another read only connection to create reports.
If that is not possible, at lease to persuade the Node client to release the lock for the other connection to read the data to create the reports.

Are any of these possible in version 0.10?

**Reply by Tishj** [reply]
I have not worked on the Node client in any big capacity, so I'm not familiar with the internals
In general though, I agree that the expected behaviour would be that `db.close()` releases the filesystem lock on the database file

You can have your read/write connection active, do reading/writing with that, close it and then establish a read-only connection to read the changes made.
Which I assume is what you are attempting to do, but the "close it" portion is not releasing the filesystem lock, right?

**Reply by Tishj** [reply]
It sounds like you should file an issue with a reproducable example here https://github.com/duckdb/duckdb-node/
On first glance though, you are closing the db, not the connection
Maybe something to experiment with, I suspect both need to be closed, or even *only* the connection needs to be closed

**Reply by dusadrian** [reply]
Thanks @Tishj, I created the [issue](https://github.com/duckdb/duckdb-node/issues/52) and referenced this discussion.
